internal/kafka/internal/services/quota/ams_quota_service.go

```diff
@@ -295,14 +297,93 @@ func (q amsQuotaService) ReserveQuota(kafka *dbapi.KafkaRequest) (string, *error
 	return resp.Subscription().ID(), nil
 }
 
-func (q amsQuotaService) DeleteQuota(subscriptionId string) *errors.ServiceError {
-	if subscriptionId == "" {
+func (q amsQuotaService) ReserveQuotaIfNotAlreadyReserved(kafka *dbapi.KafkaRequest) (string, *errors.ServiceError) {
+	instanceType, err := q.kafkaConfig.SupportedInstanceTypes.Configuration.GetKafkaInstanceTypeByID(kafka.InstanceType)
+	if err != nil {
+		return "", errors.GeneralError("failed checking current quota: %v", err)
+	}
+	kafkaBillingModel, err := instanceType.GetKafkaSupportedBillingModelByID(kafka.DesiredKafkaBillingModel)
+	if err != nil {
+		return "", errors.GeneralError("failed checking current quota: %v", err)
+	}
+
+	subscriptions, err := q.amsClient.FindSubscriptions(fmt.Sprintf("cluster_id='%s'", kafka.ClusterID))
+	if err != nil {
+		return "", errors.GeneralError("failed checking current quota: %v", err)
+	}
+	for _, subscription := range subscriptions {
+		// get the reserved resources
+		if subscription.Plan().ID() == kafkaBillingModel.AMSProduct && subscription.Status() == string(v1.SubscriptionStatusActive) {
+			reservedResouces, err := q.amsClient.GetReservedResourcesBySubscriptionID(subscription.ID())
+			if err != nil {
+				return "", errors.GeneralError("failed checking current quota: %v", err)
+			}
+
+			if arrays.AnyMatch(reservedResouces, func(rr *amsv1.ReservedResource) bool { return rr.ResourceName() == kafkaBillingModel.AMSResource }) {
+				return subscription.ID(), nil
+			}
+		}
+	}
+
+	// quota is not defined. We need to reserve a new one
+	return q.ReserveQuota(kafka)
+}
+
+func (q amsQuotaService) DeleteQuota(subscriptionID string) *errors.ServiceError {
+	if subscriptionID == "" {
 		return nil
 	}
 
-	_, err := q.amsClient.DeleteSubscription(subscriptionId)
+	_, err := q.amsClient.DeleteSubscription(subscriptionID)
 	if err != nil {
 		return errors.GeneralError("failed to delete the quota: %v", err)
 	}
 	return nil
 }
+
+// DeleteQuotaForBillingModel - Deletes the quota identified by `subscriptionID` only if it is related to received KafkaBillingModel
+func (q amsQuotaService) DeleteQuotaForBillingModel(subscriptionID string, kafkaBillingModel config.KafkaBillingModel) *errors.ServiceError {
+	if subscriptionID == "" {
+		return nil
+	}
+
+	subscription, found, err := q.GetSubscriptionByID(subscriptionID)
+	if err != nil {
+		return errors.GeneralError("failed to delete the quota: %v", err)
+	}
+
+	if !found {
+		return nil
+	}
+
+	if subscription.Plan().ID() == kafkaBillingModel.AMSProduct {
+		rr, err := q.amsClient.GetReservedResourcesBySubscriptionID(subscriptionID)
+		if err != nil {
+			return errors.GeneralError("failed to delete the quota: %v", err)
+		}
+		if arrays.AnyMatch(rr, func(resource *amsv1.ReservedResource) bool {
+			return resource.ResourceName() == kafkaBillingModel.AMSResource
+		}) {
```

Comment:
This could be rewritten to store the result of anymatch in a variable and then check the value of the variable.
I think that would improve readability as right now it is a conditional where the arugment is a method call that includes a method as parameter adding multiline and nesting into the condition itself

Reply:
done